### DISCUSSION TITLE
Redis as ChunkStorage laravel-chunk-upload?

### DISCUSSION THREAD

**Discussion by syedwhizzbridge** [pionl/laravel-chunk-upload#138] [General]
Hi Martin,

First of all, I would like to say thanks for your excellent work and secondly, I wanted to use Redis for Chunk Storage in your package can you please guide me if it's possible, Thanks

**Comment by pionl** [comment]
Hi @syedwhizzbridge 

not sure why would you need to store the chunks in redis? It is possible by implementing own save handler (probably needs some refactoring). Merging would be more trickier.

Imho, RAM is more expensive than disk (that would be used for storing temporary chunks).

Best regards,

**Comment by chandlerwilcox88** [comment]
Not sure if this is the same as @syedwhizzbridge was requesting but I would like to use Redis queues to assemble the chunked parts in local storage asynchronously.  This would greatly improve the stability and scalability of the uploads.

**Reply by pionl** [reply]
Should not to be hard to do - are you able to contribute? I can find time to guide you.

**Comment by aiavorskii** [comment]
Hi Martin!

Looking for this feature too. Want to clarify why I need it. I have AWS load balancer and multiple EC2 instances. While uploading my chunks (using dropzone) I receive them on different instances (they have separated filesystems), I also have S3 for other microservices and Redis cache. I dont' like the idea of using S3 since both chunks and file itself are needed to store temporary so Redis is best choice among those 2. So I need temp filestorage that can be accessed by all my EC2 instances

I was able to implement Redis storage by implementing `Illuminate\Contracts\Filesystem\Filesystem`

Basically I make this code put file contents under some specific key in Redis
`Storage::disk('redis')->put('test.txt', 'Hello World');`

Then I passed it to the **ChunkStorage** and found out that it wont work since it needs **getAdapter** method to be implemented and as I get it right `$this->isLocalDisk` should be true, but it is true only when it uses LocalDisk from adapter which is not implemented in my Redis storage (Filesystem)

So I decided to make my own **ChunkStorage**, but still it failed since **ParallelSave** also didn't work, cause it requires exactly lib's **ChunkStorage** passed to constructor.

Maybe you have some thoughts regarding it?

Thanks in advance and also thanks for your project!

Best regards,
Andrew

**Reply by pionl** [reply]
The solution would be have better "architecture" :) Interfaces would be best option. This should not be so hard to extract (using PHPStorm).

Not sure if others are dependent on using the 2 classes 🤔 Do you think you are able to make the PR? You can contribute the "redis" implementation to the core package if you want. 

If you do not know how, i could try to find some time.
